### DISCUSSION TITLE
이벤트와 트랜잭션 관련

### DISCUSSION THREAD

**Discussion by swonny** [woowacourse-teams/2023-3-ddang#533] [참고자료]
# 알림 트랜잭션 분리

### 알림 트랜잭션 분리

- `@Transactional(propagation = Propagation.*REQUIRES_NEW*)`
    - 알림이 실패한 경우에도 메시지 전송이 정상적으로 처리되도록 하기 위해 트랜잭션을 분리하고자 사용
    - 해당 프로퍼게이션 옵션으로 인해 트랜잭션 전파는 막을 수 있다.
        
        (새로 생성된 트랜잭션이 롤백되더라도 기존의 트랜잭션은 롤백되지 않음)
        
    - 예외 전파는 막을 수 없다. 따라서 별도의 예외 처리가 필요했다.
- `@EventListener`
    - 이벤트를 완료한 이후에 커밋이 진행된다.
    - 따라서 이벤트 리스너에서 호출한 알림 서비스 작업을 모두 마친 후 다시 이벤트 발행한 메서드로 돌아가  커밋이 되기까지 기다려야 한다.
    - DB의 write 작업은 커밋이 되었을 때 완료된다. 스프링의 `@Transactional` 어노테이션은 메서드 시작 시점에 트랜잭션을 열고 메서드가 끝나는 지점에서 커밋한다.
    - 동기 방식으로 처리한다. (default)
    - 어노테이션을 통해 비동기로 처리할 수 있다. (`@EnableAsync`)
- `@TransactionalEventListener(phase = TransactionPhase.*AFTER_COMMIT*)`
    - default `phase` 옵션이 `AFTER_COMMIT`
        - 기존 트랜잭션 커밋이 완료된 이후에만 실행된다.
        - 옵션들 : (`BEFORE_COMMIT` , `AFTER_COMMIT`, `BEFORE_ROLLBACK`, `AFTER_COMPLETION`)
    - 이벤트 리스너에서 레포지토리 저장 로직을 포함하고 있다면 커밋이 실행된 이후에 이벤트 리스너를 실행한다.
    - 이벤트 리스너에서 실행되는 메서드들은 새로운 트랜잭션이 열리는 것은 아니고, 기존에 commit된 트랜잭션에 참여한다. 트랜잭션이 커밋되더라도 트랜잭션이 사라지는 것은 아니다. 이때 commit된 transaction을 재 commit하는 것은 불가능하지만, transaction에 참여하여 읽기 작업은 가능하다. 따라서
    - 따라서 write 작업이 필요한 지점에서 새로운 트랜잭션을 열어주어야 하는데, 이때 필요한 것이  `Propagation.*REQUIRES_NEW`* 옵션이다.
    - 이벤트 리스너에서 이벤트를 처리할 때 트랜잭션이 적용되어 있어야 하는 경우 사용하면 된다.
        - 알림의 경우 메시지 전송 또는 입찰자가 발생한 상황에만(커밋까지 완료되어 트랜잭션이 적용되어 있는 경우에만) 알림 전송이 수행되어야 한다.
    
    > 이유는 @TransactionalEventListener의 경우 event publisher의 트랜잭션 안에서 동작하며, 커밋이 된 이후 추가 커밋을 허용하지 않기 때문인데요.
    > 
    
    → 이유가 궁금하다!
    
    - `@EventListener` 는 event publisher 트랜잭션 안에서 동작하지만, 메서드가 종료되는 시점에만 커밋이 발생하기 때문에 추가 커밋이 발생하지 않아 쓰기 작업이 가능했던 것이다!
    - 정말 그런지 테스트해보자!
    - event publisher 트랜잭션 안에서 동작하는 것을 어디서 확인할 수 있을까?

### 정리

- `@EventListener`
    - 이벤트 리스너가 기존 트랜잭션과 하나의 트랜잭션으로 묶인다.
- `@TransactionalEventListener`
    - 이벤트 리스너가 실행되기 전 커밋을 진행하기 때문에 (쓰기 관련) 트랜잭션은 커밋된다.
    - 따라서 이벤트 리스너가 실행되는 시점에는 이미 커밋이 되었기 때문에(트랜잭션은 종료되지 않음) 조회 작업 시 커밋한 내역을 조회할 수 있다. (테스트 필요!)
    - Q. 예외가 발생하면 기존의 커밋이 롤백되나?
- 기존의 propagation은 쓰기 작업이 없음에도 불구하고 불필요하게 transaction을 생성해주었던 것이다.

[보충 내용](https://github.com/woowacourse-teams/2023-3-ddang/discussions/736)

**Comment by apptie** [comment]
`기존의 propagation은 쓰기 작업이 없음에도 불구하고 불필요하게 transaction을 생성해주었던 것이다.` 
이 파트 관련해서 궁금한 점이 있는데 아직 보낸 알림을 DB에 저장한다거나 하는 로직은 없기 때문에 불필요하다고 말씀하신걸까요?

**Reply by swonny** [reply]
네 말씀해주신 것처럼 이해하고 작성한 게 맞습니다.
DB에 쓰기 작업이 발생하지 않는 작업이어서 새로운 트랜잭션을 열어주지 않아도 알림 전송 로직을 수행하는 데 문제가 없다고 이해했습니다!

**Reply by apptie** [reply]
이해했습니다 답변 감사합니다
